lib/src/rules/avoid_dynamic_calls.dart

```diff
@@ -0,0 +1,226 @@
+**DO** avoid method calls or property accessors on an object that is either
+explicitly or implicitly statically typed "dynamic". Dynamic calls are treated
+slightly different in every runtime environment and compiler, but most
+production modes (and even some development modes) have both compile size and
+runtime performance penalties associated with dynamic calls.
+
+Additionally, targets typed "dynamic" disabled most static analysis, meaning it
+is easier to lead to a runtime "NoSuchMethodError" or "NullError" than properly
+statically typed Dart code.
+
+Note, that despite "Function" being a type, the semantics are close to identical
+to "dynamic", and calls to an object that are typed "Function" will also trigger
+this lint.
+
+**BAD:**
+```
+void explicitDynamicType(dynamic object) {
+  print(object.foo());
+}
+
+void implicitDynamicType(object) {
+  print(object.foo());
+}
+
+abstract class SomeWrapper {
+  T doSomething<T>();
+}
+
+void inferredDynamicType(SomeWrapper wrapper) {
+  var object = wrapper.doSomething();
+  print(field.foo());
+}
+
+void functionType(Function function) {
+  function();
+}
+```
+
+**GOOD:**
+```
+void explicitType(Fooable object) {
+  object.foo();
+}
+
+void castedType(dynamic object) {
+  (object as Fooable).foo();
+}
+
+abstract class SomeWrapper {
+  T doSomething<T>();
+}
+
+void inferredType(SomeWrapper wrapper) {
+  var object = wrapper.doSomething<Fooable>();
+  object.foo();
+}
+```
+
+''';
+
+class AvoidDynamicCalls extends LintRule implements NodeLintRule {
+  AvoidDynamicCalls()
+      : super(
+          name: 'avoid_dynamic_calls',
+          description: _desc,
+          details: _details,
+          group: Group.errors,
+        );
+
+  @override
+  void registerNodeProcessors(
+    NodeLintRegistry registry,
+    LinterContext context,
+  ) {
+    assert(context != null);
+    final visitor = _Visitor(this);
+    registry
+      ..addAssignmentExpression(this, visitor)
+      ..addBinaryExpression(this, visitor)
+      ..addFunctionExpressionInvocation(this, visitor)
+      ..addIndexExpression(this, visitor)
+      ..addMethodInvocation(this, visitor)
+      ..addPostfixExpression(this, visitor)
+      ..addPrefixExpression(this, visitor)
+      ..addPrefixedIdentifier(this, visitor)
+      ..addPropertyAccess(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor<void> {
+  final LintRule rule;
+
+  _Visitor(this.rule);
+
+  bool _lintIfDynamic(Expression node) {
+    if (node?.staticType?.isDynamic == true) {
+      rule.reportLint(node);
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  void _lintIfDynamicOrFunction(Expression node) {
+    final staticType = node?.staticType;
+    if (staticType == null) {
+      return;
+    }
+    if (staticType.isDynamic || staticType.isDartCoreFunction) {
+      rule.reportLint(node);
+    }
+  }
+
+  @override
+  void visitAssignmentExpression(AssignmentExpression node) {
+    if (node.readType?.isDynamic != true) {
```

Comment:
Consider adding a comment indicating that the condition will always be false if the assignment is not a compound assignment. That isn't immediately obvious.